File: easy/3740_minimum_distance_between_three_equal_elements_i/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  int minimumDistance(std::vector<int> &nums) {
    std::unordered_map<int, std::vector<size_t>> indicesMap;

    for (size_t i = 0; i < nums.size(); ++i)
      indicesMap[nums[i]].push_back(i);

    int minDist = INT_MAX;

    for (const auto &[_, indices] : indicesMap) {
      if (indices.size() >= 3) {
        for (size_t i = 0; i < indices.size() - 2; ++i)
          minDist = std::min(minDist, 2 * (int)(indices[i + 2] - indices[i]));
      }
    }

    return minDist == INT_MAX ? -1 : minDist;
  }
};

File: easy/3740_minimum_distance_between_three_equal_elements_i/solution.py
from collections import defaultdict


class Solution:
    def minimumDistance(self, nums: list[int]) -> int:
        indices_map = defaultdict(list)

        for i, num in enumerate(nums):
            indices_map[num].append(i)

        min_dist = float("inf")

        for indices in indices_map.values():
            if len(indices) >= 3:
                for i in range(len(indices) - 2):
                    min_dist = min(min_dist, 2 * (indices[i + 2] - indices[i]))

        return min_dist if min_dist != float("inf") else -1
